src/qibo/tensorflow/custom_operators/cc/kernels/apply_gate_kernels.cu.cc

```diff
@@ -562,7 +609,25 @@ struct CollapseStateFunctor<GPUDevice, T> {
     }
 
     CollapseStateKernel<T><<<numBlocks, blockSize, 0, d.stream()>>>(
-        state, qubits, result, nsubstates, ntargets);
+        state, qubits, result, ntargets);
+
+    if (normalize) {
+      NormType *block_norms, *norms;
+      cudaMalloc((void**)&block_norms, sizeof(NormType) * blockSize);
+      cudaMalloc((void**)&norms, sizeof(NormType));
+
+      CalculateCollapsedNormKernel<T, NormType><<<1, blockSize, 0, d.stream()>>>(
+        state, block_norms, qubits, result, nstates, ntargets);
+      VectorReductionKernel<NormType><<<1, blockSize, 0, d.stream()>>>(
+        block_norms, norms);
+      NormalizeCollapsedStateKernel<T, NormType><<<numBlocks, blockSize, 0, d.stream()>>>(
+        state, norms, qubits, result, nstates, ntargets);
+
+      // if we use ``cudaFree`` here the custom operator fails if it is used
+      // twice in the same script
+      //cudaFree(block_norms);
+      //cudaFree(norms);
```

Comment:
This looks strange. You are creating pointers and not cleaning the memory so this may generate a segfault, right?

Reply:
Based on the CUDA examples that I read in order to do this, I also thought that I should do `cudaFree` in the end in order to clean the memory. When I did so, a stand-alone script that I had for checking the norm was working properly, however `test_custom_operators.py` was failing. After playing a little I realized that when running the Collapse gate twice in the same script, the norm calculation is wrong in the second time if `cudaFree` is used. This explains why tests were failing, since tests run as a single script. If I remove `cudaFree` both the stand-alone script and the tests work properly.

I just left it as a comment because I could not really explain why this happens. Is this because of Tensorflow's automatic memory management?

Reply:
The problem derives from the missing initialization of the arrays to 0. 
You have to initialize the arrays with memset:
```c++
      NormType *block_norms, *norms;
      cudaMalloc((void**)&block_norms, sizeof(NormType) * blockSize);
      cudaMemset(block_norms, 0, sizeof(NormType) * blockSize);
      cudaMalloc((void**)&norms, sizeof(NormType));
      cudaMemset(norms, 0, sizeof(NormType));
      ...
      cudaFree(block_norms);
      cudaFree(norms);
```
Anyway, I will open another PR where we replace the cuda* calls with a TF approach.

Reply:
Many thanks for this clarification. Indeed I checked using `cudaMemset` and everything works as expected with `cudaFree` enabled. It is a bit strange that it worked and tests were passing previously without explicitly initializing the arrays to 0.

I am not pushing the change since we will probably merge #275 here which fixes these issues anyway, but thanks for pointing this out.